Source/Dafny/AST/DafnyAst.cs

```diff
@@ -3393,6 +3471,7 @@ public abstract class Declaration : INamedRegion, IAttributeBearingDeclaration {
     public IToken tok;
     public IToken BodyStartTok = Token.NoToken;
     public IToken BodyEndTok = Token.NoToken;
+    public IToken FirstDeclarationToken = Token.NoToken;
```

Comment:
What's this? I would expect any AST node to be able to tell me at what position it starts and ends, and where it's trivia start and end. Why is there something specific to declarations?

Reply:
The something specific to declarations is that the leading trivia might contain some useful information.

Reply:
`FirstDeclarationToken` doesn't seem to be used in this PR except for in a test, and it doesn't make sense to me that you would store that here. If you want to know the leading trivia of the first member of a class, then you could just select the first member and get its leading trivia, right?

Reply:
Maybe I should rename it "FirstToken"? The token of a declaration is usually its symbol name. There can be a lot of tokens before such as "method", "function", "twostate"... "ghost". `FirstDeclarationToken` stores the first one of such tokens, for every declaration.
It's the only token that has the trivia that exists before a declaration, which might contain comments (I'm doing this in a follow-up PR)

Reply:
I see. I think all AST nodes in general should know their range, which would include "method", "function", etc. Declarations should also know the range of their name which we could bind to a separate field, say "NameRange". How I've also seen this is that the name is stored in a separate AST Node, so then the NameRange is simply the regular range of that child node.

Reply:
If we wanted, we could definitely add a stronger API in declarations like this:
```
class Declaration { ...
  public IToken NameRange => tok
}
```
that would consolidate the meaning of "tok", which otherwise does not carry this meaning. That would make it easier for compilers to know what the token is about. For now, I don't thing there is an urging case to do it, but I might be biased as usual.

In a follow-up PR, all AST nodes will be truly aware of their entire range. For that, they will need to have ownership of tokens (and their trivia optionally), but being able to access their children's token in the case of prefix, infix  postfix operations or parentheses calls.

Reply:
> In a follow-up PR, all AST nodes will be truly aware of their entire range.

Would that make `FirstDeclarationToken` obsolete then?

Reply:
No, it won't be obsolete, I will just add a new field `FinalDeclarationToken`.
I could also rename them StartToken and EndToken if you preferred.

Reply:
> No, it won't be obsolete, I will just add a new field FinalDeclarationToken.
I could also rename them StartToken and EndToken if you preferred.

Do we agree on what the end-state should look like? I imagine this:

```
record TokenRange(start: Token, end: Token);
record ASTNode(range: TokenRange);
record Declaration(nameRange: TokenRange) : ASTNode
```

Given that, I don't see how introducing `FirstDeclarationToken` moves toward that, since then there won't be a need for the `FirstDeclarationToken ` and `FinalDeclarationToken` fields in `Declaration` right?

Reply:
For a single node, we want and will want to access tokens for different reasons:
- To access the token that is just before the official whitespace where we expect documentation.
- To access certain tokens that delimits an inner region, not the entire node (e.g. the body of a method)
- To access certain tokens that indicate whether some keyword was added or not (existing behavior, see extends, refines, etc.)
- To access the first token of a declaration
- To access the final token of a declaration, to know where the printer should stop
- To access the range, with the start token and the end token
- To access individual owned tokens to be able to print them later

We could define a single field "Range" as you state, and tell our users that if they want the first token, they have to use `RangeToken.Start` instead of `FirstDeclarationToken`. But I developed this PR based on needs: Our users probably don't care about the final token of a declaration. It's mostly an internal thing for the pretty-printer down the road, and possibly for refactoring. So why should we create a TokenRange so early? Compiler implementers, will care more about the first token than the range because the first token that is likely to contain docstring.
In this PR, we don't need the final token at all.